modules/channel-messenger/src/backend/messenger.ts

```diff
@@ -0,0 +1,335 @@
+import crypto from 'crypto'
+import { Router } from 'express'
+import _ from 'lodash'
+
+import { Config } from '../config'
+
+const debug = DEBUG('channel-messenger')
+const debugMessages = debug.sub('messages')
+const debugHttp = debug.sub('http')
+const debugWebhook = debugHttp.sub('webhook')
+const debugHttpOut = debugHttp.sub('out')
+
+const outgoingTypes = ['text', 'typing', 'login_prompt', 'carousel']
+type MessengerAction = 'typing_on' | 'typing_off' | 'mark_seen'
+
+type MountedBot = { pageId: string; botId: string; client: MessengerClient }
+
+export class MessengerService {
+  private readonly http = axios.create({ baseURL: 'https://graph.facebook.com/v3.2/me' })
+  private mountedBots: MountedBot[] = []
+  private router: Router & sdk.http.RouterExtension
+  private appSecret: string
+
+  constructor(private bp: typeof sdk) {}
+
+  async initialize() {
+    const config = (await this.bp.config.getModuleConfig('channel-messenger')) as Config
+
+    if (!config.verifyToken || config.verifyToken.length < 1) {
+      throw new Error('You need to set a non-empty value for "verifyToken" in the *global* messenger config')
+    }
+
+    if (!config.appSecret || config.appSecret.length < 1) {
+      throw new Error(`You need to provide your app's App Secret in the *global* messenger config`)
+    }
+
+    this.appSecret = config.appSecret
+
+    this.router = this.bp.http.createRouterForBot('channel-messenger', {
+      checkAuthentication: false,
+      enableJsonBodyParser: false // we use our custom json body parser instead, see below
+    })
+
+    this.router.getPublicPath().then(publicPath => {
+      if (publicPath.indexOf('https://') !== 0) {
+        this.bp.logger.warn('Messenger requires HTTPS to be setup to work properly. See EXTERNAL_URL botpress config.')
+      }
+
+      this.bp.logger.info(`Messenger Webhook URL is ${publicPath.replace('BOT_ID', '___')}/webhook`)
+    })
+
+    this.router.use(
+      bodyParser.json({
+        verify: this._verifySignature.bind(this)
+      })
+    )
+
+    this.router.get('/webhook', this._setupWebhook.bind(this))
+    this.router.post('/webhook', this._handleIncomingMessage.bind(this))
+
+    this.bp.events.registerMiddleware({
+      description: 'Sends outgoing messages for the messenger channel',
+      direction: 'outgoing',
+      handler: this._handleOutgoingEvent.bind(this),
+      name: 'messenger.sendMessages',
+      order: 200
+    })
+  }
+
+  async mountBot(botId: string) {
+    const config = (await this.bp.config.getModuleConfigForBot('channel-messenger', botId)) as Config
+    if (config.enabled) {
+      if (!config.accessToken) {
+        return this.bp.logger
+          .forBot(botId)
+          .error('You need to configure an Access Token to enable it. Messenger Channel is disabled for this bot.')
+      }
+
+      const { data } = await this.http.get('/', { params: { access_token: config.accessToken } })
+
+      if (!data && !data.id) {
```

Comment:
Will fail with a null pointer if data is null

Reply:
good catch, meant to write || ;)